### ISSUE TITLE
[bug] Inconsistent checking of a refined data type's field assertion related to `Integer`

### ISSUE THREAD

**Issue by plredmond** [ucsd-progsys/liquidhaskell#1758]
**Problem:**
A refined data type's field assertion is not consistently checked.
```haskell
-- (1)

{-@
type Nat = {v:Int | 0 <= v} @-}
type Nat = Int

{-@
newtype Pos = Pos Nat @-}
newtype Pos = Pos Nat

{-@
bad :: Pos @-}
bad :: Pos
bad = Pos (-1) -- ERROR! (as expected)
```
This code (1) is correctly rejected by LiquidHaskell. If you `s/Int/Integer` it (2) is accepted.
```haskell
-- (2)

{-@
type Nat = {v:Integer | 0 <= v} @-}
type Nat = Integer

{-@
newtype Pos = Pos Nat @-}
newtype Pos = Pos Nat

{-@
bad :: Pos @-}
bad :: Pos
bad = Pos (-1)

-- This code passes (but probably shouldn't)
```
**Expected:**
The version using `Integer` (2) is rejected by LiquidHaskell.

**Actual:**
LH accepts the code (2).

<details><summary><strong>Workaround:</strong>

Inlining the refinement to the refined data type fixes the problem (2) and LiquidHaskell correctly rejects the code (3).

Click to expand.
</summary>

```haskell
-- (3)

{-@
newtype Pos = Pos { field :: {v:Integer | 0 <= v} } @-}
newtype Pos = Pos { field :: Integer }

{-@
bad :: Pos @-}
bad :: Pos
bad = Pos (-1) -- ERROR! (as expected)
```

</details>

---

<details><summary><strong>Possibly related bug:</strong>

Changing the broken example (2) to add a field name/getter to the newtype (4) causes a possibly related bug where LiquidHaskell rejects the code with a "does not refine Haskell type" about the field name/getter. This error is also connected to the use of `Integer` and it resolves when `Int` is used.

Click to expand.
</summary>

```haskell
-- (4)

{-@
type Nat = {v:Integer | 0 <= v} @-}
type Nat = Integer

{-@
newtype Pos = Pos { unpos :: Nat } @-} -- ERROR! (unexpectedly..)
newtype Pos = Pos { unpos :: Nat }

{-@
bad :: Pos @-}
bad :: Pos
bad = Pos (-1)
```
```
lib/Causal/VectorClock.hs:12:9-12: error:
    • Specified type does not refine Haskell type for `Causal.VectorClock.unpos` (Checked)
```
Changing this back `s/Integer/Int` resolves the issue such that the code is rejected correctly as in the original (1). Since this problem also relies to the use of `Integer` I think this may be have a related underlying cause.

</details>

---

**Version:**
`f2871af95 - Merge pull request #1736 from ucsd-progsys/lts-16`

**Comment by ranjitjhala** [comment]
AFAICT the `Integer` case doesn't actually work -- it creates a crashing malformed constraint?

However, if you change `newtype` to `data` its rejected as expected.

This is, unfortunately because of a long standing issue with `newtype` -- namely that there are no "constructors" 
and instead everything happens with "casts".

In principle though we ought to be able to deal with the cast situation ...

**Comment by plredmond** [comment]
Ah, ok, if this is a known issue related to the treatment of `newtypes` in LH, is there an issue I should link to and then close this as a duplicate?

**Comment by plredmond** [comment]
> AFAICT the Integer case doesn't actually work -- it creates a crashing malformed constraint?

Oops! My mistake. I've been using `ghci` because the issue where LH errors weren't reported (#1711) seemed to have been fixed. This bug is an instance of #1711, then? Shall I close this?

**Comment by ranjitjhala** [comment]
Yes, this is a super old issue e.g. see #708

**Comment by plredmond** [comment]
Closed as a duplicate of #708 and an instance of #1711.
